Set knowledge base slots (#4210)
RasaHQ/rasa: docs/core/knowledge-bases.rst

:desc: Leverage information from knowledge bases inside conversations using ActionQueryKnowledgeBase
       in open source bot framework Rasa.

.. _knowledge_bases:

Knowledge Base Actions
======================

.. edit-link::

.. warning::

>    This feature is experimental.

Reviewer: ```suggestion
   This feature is experimental. We introduce experimental features to get feedback on implementations, so do please try it out. However, the functionality might be changed or removed in the future. If you have feedback (positive or negative) please comment in the `community forum <https://forum.rasa.com>`_
```

.. contents::
   :local:

> A lot of users want to obtain detailed information about certain objects, such as restaurants or hotels, during a conversation.

Reviewer: I think we can make this easier to understand by motivating with an example. Otherwise this sounds very abstract

They want to obtain details, such as, if an restaurant has outside seating or how expensive the restaurant is.
In order to answer those user requests domain knowledge is needed.
Hard-coding the information would not help as the information are subject to change.
Additionally, users do not only refer to objects by their names, but also use terms, such as "the first one" or "that
restaurant", to refer to a specific object.
Objects need to be recognised and reused at a later point in the conversation.

To handle the above challenges, we recommend that you create a custom action that inherits from ``ActionQueryKnowledgeBase``.
This is a single actions which contains the logic to query a knowledge base for objects and their attributes.
The action is also able to resolve certain mention of objects, such as ``the first one`` or ``that restaurant``.
You can find a complete example in ``examples/knowledge_base_bot``.

.. note::
   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
   ``ActionQueryKnowledgeBase``.

> Create a KnowledgeBase

Reviewer: ```suggestion
Create a Knowledge Base
```

            {
                "id": 0,
                "name": "Hilton",
                "price-range": "expensive",
                "breakfast-included": true,
                "city": "Berlin",
                "free-wifi": true,
                "star-rating": 5,
                "swimming-pool": true
            },
            {
                "id": 1,
                "name": "Hilton",
                "price-range": "expensive",
                "breakfast-included": true,
                "city": "Frankfurt am Main",
                "free-wifi": true,
                "star-rating": 4,
                "swimming-pool": false
            },
            {
                "id": 2,
                "name": "B&B",
                "price-range": "mid-range",
                "breakfast-included": false,
                "city": "Berlin",
                "free-wifi": false,
                "star-rating": 1,
                "swimming-pool": false
            },
        ]
    }

The above json file contains data about restaurants and hotels.
The json structure should contain a key for every object type, i.e. "restaurant" and "hotel".
Every object type maps to a list of objects.

> Once the data are defined in a json file, called, for example, ``data.json``, you can create your

Reviewer: did we get rid of the `schema` part? if so that's a nice simplification

Author: yes, split it up into separate methods and set default values

``InMemoryKnowledgeBase``:

.. code-block:: python

    knowledge_base = InMemoryKnowledgeBase.load("data.json")

> Customize your InMemoryKnowledgeBase

Reviewer: I feel like this section contains a lot of advanced details that most people don't need, correct? If I were looking to get started with this feature I would find all of this intimidating

Author: depends. if the user is using the default values, e.g. every object has an id and a name field, he does not need to do anything. if not then he should overwrite those methods.

Reviewer: ok. does it work to then mention that every item in the KB should have a `name` and `id` field, and if not, they should read about customizing below?

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The class ``InMemoryKnowledgeBase`` inherits ``KnowledgeBase``.
You can customize your ``InMemoryKnowledgeBase`` by overwriting the following functions:

- ``get_key_attribute_of_object``: To keep track of what object the user was talking about last, we store the value
  of the key attribute in a specific slot. Every object should have a key attribute that is unique, i.e.

>   similar to the primary key in a relation database. Per default the name of the key attribute for every object type

Reviewer: ```suggestion
  similar to the primary key in a relation database. By default the name of the key attribute for every object type
```

  is set to "id". You can overwrite the name of the key attribute for a specific object type by calling
  ``set_key_attribute_of_object()``.

> - ``get_representation_function_of_object``: This methods returns a lambda function that maps an object from the

Reviewer: this is a difficult sentence (part of a difficult paragraph). Let's split it up into smaller sentences and give quick examples

>   knowledge base to a string representation. This function is used whenever an object is outputted to the user.

Reviewer: 'outputted' sounds awkward. what are we actually trying to say?

Author: bot is referring to an object. rephrased it.

>   Per default the lambda function is set to ``lambda obj: obj["name"]``. So, it returns the value of the attribute

Reviewer: ```suggestion
  By default the lambda function is set to ``lambda obj: obj["name"]``. So, it returns the value of the attribute
```

changed to "By default" - "Per" is also correct but less common

  "name" of the object. If your object does not have an attribute "name", or the "name" of an object might be
  ambiguous, you should set a new lambda function for that object type by calling
  ``set_representation_function_of_object()``.
- ``set_ordinal_mention_mapping``: The ordinal mention mapping is needed to resolve an ordinal mention to an object
  in a list. For example, if the bot listed a few restaurants in Berlin, and the user then asked "Does the second one
  have outside seating?", you need to resolve "second one" to the correct object the bot listed before. Per
  default the ordinal mention mapping looks like this:

  .. code-block:: python

      {
          "1": lambda l: l[0],
          "2": lambda l: l[1],
          "3": lambda l: l[2],
          "4": lambda l: l[3],
          "5": lambda l: l[4],
          "6": lambda l: l[5],
          "7": lambda l: l[6],
          "8": lambda l: l[7],
          "9": lambda l: l[8],
          "10": lambda l: l[9],
          "ANY": lambda l: random.choice(list),
          "LAST": lambda l: l[-1],
      }

  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.

>   More on the usage can be found in section :ref:`resolve_mentions`.

Reviewer: Avoid passive voice like "More on the usage can be found". Instead say:
"If you want to learn more about X, go to ..."

Creating your own Knowledge Base Implementation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you have more data or if you want to use a more complex data structure that, for example, involves relations between
different objects, you can also create your own knowledge base implementation.
Just inherit ``KnowledgeBase`` and implement the methods ``get_objects()``, ``get_object()``, and
``get_attributes_of_object()``.
You can also customize your knowledge base further, for example, by adapting the methods mentioned in the previous
section.

.. note::
   We wrote a `blog post <https://blog.rasa.com/set-up-a-knowledge-base-to-encode-domain-knowledge-for-rasa/>`_
   that explains how you can set up your own knowledge base.

> Create an ActionQueryKnowledgeBase

Reviewer: I'd move this further down, below the NLU data description. If people want to set up a "proper" KB they can skip to that section

Berlin.
In order to filter the objects in the knowledge base, you need to mark "Italian" and "Berlin" as entities.
E.g. ``What [Italian](cuisine) [restaurant](object_type) options in [Berlin](city) do I have?``.
The names of the attributes, e.g. "cuisine" and "city", should be equal to the ones used in the knowledge base.
You also need to add those entities as entities and slots in the domain file.
If the NER detects those attributes in the request of the user, the action will use those for filter the
restaurants found in the knowledge base.

Once the bot retrieved some entities from the knowledge base, it will response to the user with

    `Found the following objects of type 'restaurant':`
    `1: I due forni`
    `2: PastaBar`
    `3: Berlin Burrito Company`

Or if no entities could be found

    `I could not find any objects of type 'restaurant'.`

If you want to change the utterance of the bot, you can overwrite the method ``utter_objects()`` in your action.

Query the Knowledge Base for an Attribute of an Object
------------------------------------------------------

> To obtain the value of an attribute for a specific object from the knowledge base, the action needs to know the object

Reviewer: This is super difficult language: "To obtain the value of an attribute for a specific object" - I think we can say this in plain english by usign an example

:desc: Leverage information from knowledge bases inside conversations using ActionQueryKnowledgeBase
       in open source bot framework Rasa.

.. _knowledge_bases:

Knowledge Base Actions
======================

.. edit-link::

.. warning::
   This feature is experimental.
   We introduce experimental features to get feedback on implementations, so do please try it out.
   However, the functionality might be changed or removed in the future.
   If you have feedback (positive or negative) please comment in the `community forum <https://forum.rasa.com>`_.

.. contents::
   :local:

Knowledge base actions enable you to handle the following kind of dialogues:

.. image:: ../_static/images/knowledge-base-example.png

As you can see users might want to obtain detailed information about certain objects, such as restaurants or hotels,
during a conversation.
They want to obtain details, such as, if an restaurant has outside seating or how expensive the restaurant is.
In order to answer those user requests domain knowledge is needed.
Hard-coding the information would not help as the information are subject to change.

> Additionally, users do not only refer to objects by their names, but also use terms, such as "the first one" or "it",

Reviewer: I think this is the key point that the KB helps with so should be mentioned first.

Author: not really the KB, but more the smart usage of slots

to refer to a specific restaurant.

> Restaurants need to be recognised and reused at a later point in the conversation.

Reviewer: I'm not sure what this sentence is telling me

To handle the above challenges, we recommend that you create a custom action that inherits from ``ActionQueryKnowledgeBase``.
This is a single actions which contains the logic to query a knowledge base for objects and their attributes.

> The action is also able to resolve certain mention of objects, such as ``the first one`` or ``that restaurant``.

Reviewer: ```suggestion
When a restaurant is mentioned indirectly, for example using a phrase like ``the first one`` or ``that restaurant``, this action is able to figure out which restaurant the user is referring to.
```

You can find a complete example in ``examples/knowledge_base_bot``.

.. note::
   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
   ``ActionQueryKnowledgeBase``.

Create a Knowledge Base
-----------------------

The data you will use to answer the user's request comes from a knowledge base.
A knowledge base can be used to store complex data structures.

> If you just have some data points that fit into memory, you can use our ``InMemoryKnowledgeBase`` implementation.

Reviewer: ```suggestion
We suggest you get started by using the ``InMemoryKnowledgeBase``. Once you want to start working with a large amount of data, you can switch to a :ref:`custom-knowledge-base` .
```

  ambiguous, you should set a new lambda function for that object type by calling
  ``set_representation_function_of_object()``.
- ``set_ordinal_mention_mapping``: The ordinal mention mapping is needed to resolve an ordinal mention to an object
  in a list. For example, if the bot listed a few restaurants in Berlin, and the user then asked "Does the second one
  have outside seating?", you need to resolve "second one" to the correct object the bot listed before. Per
  default the ordinal mention mapping looks like this:

  .. code-block:: python

      {
          "1": lambda l: l[0],
          "2": lambda l: l[1],
          "3": lambda l: l[2],
          "4": lambda l: l[3],
          "5": lambda l: l[4],
          "6": lambda l: l[5],
          "7": lambda l: l[6],
          "8": lambda l: l[7],
          "9": lambda l: l[8],
          "10": lambda l: l[9],
          "ANY": lambda l: random.choice(list),
          "LAST": lambda l: l[-1],
      }

  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
  If you want to learn more about the usage of the mapping, go to section :ref:`resolve_mentions`.

> Defining the NLU Data

Reviewer: Let's start this section with a brief summary:

- we are going to introduce a new intent, ``query_knowledge_base``
- we are going to annotate `mention` entities so that our model detects indirect mentions of entities like 'the first one'
- we will use synonyms extensively

attribute of an object.
The ``ActionQueryKnowledgeBase`` can handle both of those requests.
Other requests, such as comparison between objects, are currently not supported.

Another thing you may have noticed is, that we marked different kind of entities in the NLU data.
If you want to use ``ActionQueryKnowledgeBase``, you need to specify the following entities:

- ``object_type``: Whenever the user is talking about a specific object type from your knowledge base, the type should
  be marked as entity in our NLU data. Use :ref:`entity_synonyms` to map, for example, "restaurants" to the correct
  object type listed in the knowledge base, e.g. "restaurant".
- ``mention``: If the user refers to an object via "the first one", "that one", or "it", you should mark those terms
  as ``mention``. We also use :ref:`entity_synonyms` to map some of the mentions to symbols. More on that in
  :ref:`resolve_mentions`.
- ``attribute``: All attribute names defined in your knowledge base should be marked as ``attribute`` in the NLU data.
  Again, use :ref:`entity_synonyms` to map variations of an attribute name to the one used in the knowledge base.

> Don't forget to add those entities to your domain file (as entities and slots).

Reviewer: ```suggestion
Remember to add those entities to your domain file (as entities and slots).
```

Reviewer: do they need to be slots? should the be unfeaturized? let's include a snippet (possibly from the example)

> 

Reviewer: ```suggestion

.. _custom_knowledge_base:

```

:desc: Leverage information from knowledge bases inside conversations using ActionQueryKnowledgeBase
       in open source bot framework Rasa.

.. _knowledge_base_actions:

Knowledge Base Actions
======================

.. edit-link::

.. warning::
   This feature is experimental.
   We introduce experimental features to get feedback on implementations, so do please try it out.
   However, the functionality might be changed or removed in the future.

>    If you have feedback (positive or negative) please comment in the `community forum <https://forum.rasa.com>`_.

Author: No, but we could create one. But maybe first release the feature, then create a forum post so that we can link to the documentation, and finally update the documentation to link to the correct forum post.

>    If you have feedback (positive or negative) please comment in the `community forum <https://forum.rasa.com>`_.

Reviewer 2: Yes, I suppose that's probably the best way to do it.

.. contents::
   :local:

Knowledge base actions enable you to handle the following kind of dialogues:

.. image:: ../_static/images/knowledge-base-example.png

As you can see users might want to obtain detailed information about certain objects, such as restaurants or hotels,
during a conversation.

> They want to obtain details, such as, if an restaurant has outside seating or how expensive the restaurant is.

Reviewer 3: ```suggestion
They want to obtain details, such as, if a restaurant has outside seating or how expensive the restaurant is.
```

In order to answer those user requests domain knowledge is needed.
Hard-coding the information would not help as the information are subject to change.
Additionally, users do not only refer to objects by their names, but also use terms, such as "the first one" or "it",
to refer to a specific restaurant.
We need to keep track of what the user spoke about in order to resolve mentions, such "the first one" or "it", to
the correct object.

To handle the above challenges, we recommend that you create a custom action that inherits from
``ActionQueryKnowledgeBase``.
This is a single actions which contains the logic to query a knowledge base for objects and their attributes.
When a restaurant is mentioned indirectly, for example using a phrase like ``the first one`` or ``that restaurant``,
this action is able to figure out which restaurant the user is referring to.
You can find a complete example in ``examples/knowledge_base_bot``.

.. note::
   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind

>    ``ActionQueryKnowledgeBase``.

Reviewer 2: I've commented later on a proposed structure, where this tutorial link would be included much later in the customization section.

                "star-rating": 4,
                "swimming-pool": false
            },
            {
                "id": 2,
                "name": "B&B",
                "price-range": "mid-range",
                "breakfast-included": false,
                "city": "Berlin",
                "free-wifi": false,
                "star-rating": 1,
                "swimming-pool": false
            },
        ]
    }

The above json file contains data about restaurants and hotels.
The json structure should contain a key for every object type, i.e. "restaurant" and "hotel".
Every object type maps to a list of objects.

Once the data are defined in a json file, called, for example, ``data.json``, you can create your
``InMemoryKnowledgeBase``:

.. code-block:: python

    knowledge_base = InMemoryKnowledgeBase("data.json")

Every object in your knowledge base should have "name" and "id" field.
If that is not the case, please read the following section :ref:`customize_in_memory_knowledge_base`.

.. _customize_in_memory_knowledge_base:

Customize your InMemoryKnowledgeBase
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The class ``InMemoryKnowledgeBase`` inherits ``KnowledgeBase``.

> You can customize your ``InMemoryKnowledgeBase`` by overwriting the following functions:

Author: Two ways of customizations:
1. If you don't follow all assumptions we made, you need to adapt the `InMemoryKnowledgeBase`. E.g. not all your objects have an `id` field or a `name` field.
2. If you want to use a graph database or any other database in general, you need to create a new class that inherits `KnowledgeBase`.

  ambiguous, you should set a new lambda function for that object type by calling
  ``set_representation_function_of_object()``.
- ``set_ordinal_mention_mapping``: The ordinal mention mapping is needed to resolve an ordinal mention to an object
  in a list. For example, if the bot listed a few restaurants in Berlin, and the user then asked "Does the second one
  have outside seating?", you need to resolve "second one" to the correct object the bot listed before. Per
  default the ordinal mention mapping looks like this:

  .. code-block:: python

      {
          "1": lambda l: l[0],
          "2": lambda l: l[1],
          "3": lambda l: l[2],
          "4": lambda l: l[3],
          "5": lambda l: l[4],
          "6": lambda l: l[5],
          "7": lambda l: l[6],
          "8": lambda l: l[7],
          "9": lambda l: l[8],
          "10": lambda l: l[9],
          "ANY": lambda l: random.choice(list),
          "LAST": lambda l: l[-1],
      }

  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
  If you want to learn more about the usage of the mapping, go to section :ref:`resolve_mentions`.

> Creating your own Knowledge Base Implementation

Author: I like the structure 👍 Thanks! Will update the documentation.

If you want to use ``ActionQueryKnowledgeBase``, you need to specify the following entities:

- ``object_type``: Whenever the user is talking about a specific object type from your knowledge base, the type should
  be marked as entity in our NLU data. Use :ref:`entity_synonyms` to map, for example, "restaurants" to the correct
  object type listed in the knowledge base, e.g. "restaurant".
- ``mention``: If the user refers to an object via "the first one", "that one", or "it", you should mark those terms
  as ``mention``. We also use :ref:`entity_synonyms` to map some of the mentions to symbols. You can learn about that
  in section :ref:`resolve_mentions`.
- ``attribute``: All attribute names defined in your knowledge base should be identified as ``attribute`` in the
  NLU data. Again, use :ref:`entity_synonyms` to map variations of an attribute name to the one used in the
  knowledge base.

Remember to add those entities to your domain file (as entities and slots):

.. code-block:: md

    entities:
      - object_type
      - mention
      - attribute

    slots:
      object_type:
        type: text
      mention:
        type: text
      attribute:

>         type: text

Author: I think they should be featurized as they indicate the query knowledge base intent quite well.

>         type: text

Reviewer 2: But it is the entities featurization that does that, not the slot featurization right? 

If we keep them as text, maybe we should add a note to remember to put the slot events in the stories (but then there would be different slots being filled depending on what the input is that cause the same action -- which leads to more stories for all of the different combos, right?

>         type: text

Author: not sure about entities featurization vs slot featurization... will figure it out. Would be helpful if we could keep them as unfeaturized slots as that would simplify the stories for sure.

>         type: text

Reviewer 2: @Ghostvv might have insight on this

>         type: text

Author: I guess, you are right. Entities are added as features to the intent classification anyway. So, nothing needs to be done here. If slot are featurized, they are used as features for predicting the next action. That is actually not needed. So, we can change the types to unfeaturized. (Finally, I understood it 😃 )

``KnowledgeBase`` implementation (see :ref:`customize_in_memory_knowledge_base`).
As the ordinal mention mapping does not, for example, include an entry for "the first one".
It is important that you use :ref:`entity_synonyms` to map "the first one" in your NLU data to "1".
For example,

.. code-block:: md

    Does the [first one](mention:1) have [outside seating](attribute:outside-seating)?

maps "first one" via a synonym to "1".
The NER detects "first one" as ``mention`` entity, but puts "1" into the ``mention`` slot.
Thus, our action can take the ``mention`` slot together with the ordinal mention mapping to resolve "first one" to
the actual object "I due forni".

Other Mentions
~~~~~~~~~~~~~~
Take a look at the following conversation:

- User: `What is the cuisine of PastaBar?`
- Bot: `PastaBar has an Italian cuisine.`
- User: `Does it have wifi?`
- Bot: `Yes.`
- User: `Can you give me an address?`

In the second utterance of the user, the user refers to "PastaBar" by the word "it".
If the NER detected "it" as the entity ``mention``, the knowledge base action would resolve it to the last mentioned

> object in the conversation, e.g. "PastaBar".

Author: What do you mean? The NER should detect "it" as `mention`. Otherwise the action might not be able to resolve it to the last mentioned object (in case `use_last_object_mention` is set to False).

> object in the conversation, e.g. "PastaBar".

Reviewer 2: Ah okay. I just mean when creating the NLU data, then we should also be tagging `it` as a mention? e.g. in L160:
```
what [cuisine](attribute) is it?
```
we don't tag `it` as a mention

> object in the conversation, e.g. "PastaBar".

Author: good catch! will add it.

> object in the conversation, e.g. "PastaBar".

Reviewer 2: makes sense! another couple of examples of mentions like `it` and `that one` would be helpful

  in a list. For example, if the bot listed a few restaurants in Berlin, and the user then asked

    `Does the second one have outside seating?`

  you need to resolve "second one" to the correct object the bot listed before. Per
  default the ordinal mention mapping looks like this:

  .. code-block:: python

      {
          "1": lambda l: l[0],
          "2": lambda l: l[1],
          "3": lambda l: l[2],
          "4": lambda l: l[3],
          "5": lambda l: l[4],
          "6": lambda l: l[5],
          "7": lambda l: l[6],
          "8": lambda l: l[7],
          "9": lambda l: l[8],
          "10": lambda l: l[9],
          "ANY": lambda l: random.choice(list),
          "LAST": lambda l: l[-1],
      }

  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
  If you want to learn more about the usage of the mapping, go to section :ref:`resolve_mentions`.

Customize your `ActionQueryKnowledgeBase`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can overwrite the following two functions of `ActionQueryKnowledgeBase`:

- ``utter_objects()``
- ``utter_attribute_value()``

> ``utter_objects()`` is used when the user requested the bot to list some objects.

Author: If I want to overwrite those functions, I would need to create a new knowledge base. I don't want that. I added doc strings to all methods and the `InMemoryKnowledgeBase` is basically an example implementation of the `KnowledgeBase` class. I think that is enough.
In the knowledgebasebot example I'm overwriting one of the functions. So, it should be clear how to do it.

> ``utter_objects()`` is used when the user requested the bot to list some objects.

Reviewer 2: > If I want to overwrite those functions, I would need to create a new knowledge base. I don't want that. I added doc strings to all methods and the InMemoryKnowledgeBase is basically an example implementation of the KnowledgeBase class. I think that is enough.

Got it. Maybe we should link to the code where those methods are found

> In the knowledgebasebot example I'm overwriting one of the functions. So, it should be clear how to do it.

Similarly, reference that there is an example of this in the knowledgebasebot example since we're pretty far down on the page by now

If you want to change the utterance of the bot, you can overwrite the method ``utter_attribute_value()``.

.. note::
   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
   ``ActionQueryKnowledgeBase``.

.. _custom_knowledge_base:

Creating your own Knowledge Base
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you have more data or if you want to use a more complex data structure that, for example, involves relations between
different objects, you can create your own knowledge base implementation.
Just inherit ``KnowledgeBase`` and implement the methods ``get_objects()``, ``get_object()``, and
``get_attributes_of_object()``.
You can also customize your knowledge base further, for example, by adapting the methods mentioned in the section
:ref:`customize_in_memory_knowledge_base`.

.. note::
   We wrote a `blog post <https://blog.rasa.com/set-up-a-knowledge-base-to-encode-domain-knowledge-for-rasa/>`_
   that explains how you can set up your own knowledge base.

> Creating your own Knowledge Base Actions

Reviewer 2: Should probably go above creating your own knowledge base -- feel like the level of customisation would make most sense as

1. Customize actionqueryknowledgebase 
2. Create your own knowledge base action 
3. Customize the in memory knowledge base 
4. Create your own knowledge base

    - what is with [I due forni](restaurant)?
    - can you tell me the [price range](attribute) of [that restaurant](mention)?
    - what [cuisine](attribute) do [they](mention) have?
     ...

The above example just shows examples related to the restaurant domain.
You should add examples for every object type that exists in your knowledge base.

All user requests can be divided into two categories:
(1) The user wants to obtain a list of objects of a specific type or (2) the user wants to know about a certain
attribute of an object.
The ``ActionQueryKnowledgeBase`` can handle both of those requests.
Other requests, such as comparison between objects, are currently not supported.

Another thing you may have noticed is, that we marked different kind of entities in the NLU data.
If you want to use ``ActionQueryKnowledgeBase``, you need to specify the following entities:

- ``object_type``: Whenever the user is talking about a specific object type from your knowledge base, the type should
  be marked as entity in our NLU data. Use :ref:`entity_synonyms` to map, for example, "restaurants" to the correct

>   object type listed in the knowledge base, e.g. "restaurant".

Reviewer 2: Do they also need to add the 
```
##synonym: 1
- the first one

##synonym: outside-seating
- outside seating
```
etc. to the NLU data?

Author: You can, but you don't need to. The bot is working just fine without defining it. As explained in the `entity_synonyms` section, two ways exists to express synonyms. I guess we can leave it as it is.

Reviewer 2: Got it. Just didn't know exactly how synonyms work 😃
